### ISSUE TITLE
GOAWAY for graceful shutdown 

### ISSUE THREAD

**Issue by benjchristensen** [ReactiveSocket/reactivesocket#59]
Do we need something similar to HTTP/2 GOAWAY?

The use case I have in mind is a server shedding load and disconnecting a connection and letting a client know. 

We don't currently have a connection-level error frame. 

Example use case from HTTP/2 spec:

> Servers are encouraged to maintain open connections for as long as possible but are permitted to terminate idle connections if necessary. When either endpoint chooses to close the transport-layer TCP connection, the terminating endpoint SHOULD first send a GOAWAY (Section 6.8) frame so that both endpoints can reliably determine whether previously sent frames have been processed and gracefully complete or terminate any necessary remaining tasks.

**Comment by stevegury** [comment]
Yes, I just mentioned that in #25, a server needs an explicit way to reject one individual message.
The client has to know that the server didn't process the message, and know that it's safe to retry it (even with non-idempotent messages).

**Comment by benjchristensen** [comment]
The GOAWAY frame though is not for an individual message, it is for the entire connection. 

This is similar to what is being discussed in https://github.com/ReactiveSocket/reactivesocket/issues/50 where we need a connection-level error frame. 

At the individual stream level, we can already use the ERROR frame to reject an individual request.

**Comment by NiteshKant** [comment]
GOAWAY frame is useful for a few things in HTTP/2:

* Stop accepting new streams but finish processing of existing streams.
* Send the last received stream ID before the GOAWAY frame was sent. This makes the client gracefully complete the processing of the existing streams but not send or expect to receive any data for the streams that are initiated after the last received stream by the peer.
* The peer sending the GOAWAY frame need not send any data for the streams initiated post sending GOAWAY frame. This will save sending the ERROR frame for those streams.

If we are not doing these and just indicating error in the frame and close the connection immediately. I think it will have very limited value as opposed to just closing the connection. Graceful shutdown is an important aspect of GOAWAY frame IMO.

**Comment by tmontgomery** [comment]
@NiteshKant has a good point. I'm not sure that graceful shutdown is needed with the LEASE semantic in place. Just send a null LEASE that stops new requests, then send COMPLETEs or ERRORs (for subscriptions).

**Comment by tmontgomery** [comment]
Mentioned in #53 . Repeating here.

We could:

- Define ERROR to be allowed on Stream ID 0.
- Remove SETUP_ERROR as a separate frame type.
- Use ERROR on Stream ID 0 to mean connection level error with connection teardown (?).

Whether seen during SETUP or during operation, ERROR on Stream ID 0 is not necessary to distinguish between. The operation is the same. Inform application, shutdown, etc.

**Comment by NiteshKant** [comment]
>>> I'm not sure that graceful shutdown is needed with the LEASE semantic in place. Just send a null LEASE that stops new requests, then send COMPLETEs or ERRORs (for subscriptions).

Me @stevegury and @benjchristensen were discussing this offline. Lease 0 isn't exactly the same as the GOAWAY frame as the GOAWAY frame also contain the last received Stream ID which is useful for the peer to not send any more data for the streams that would not be processed in addition to not sending any more requests.

**Comment by stevegury** [comment]
My understanding of GOAWAY is that it's equivalent to a LEASE 0 + a "reject" of all the outstanding requests (in a more compact and efficient way).

**Comment by NiteshKant** [comment]
@stevegury "all the outstanding requests" has an implicit race condition as described [in the HTTP/2 spec](https://http2.github.io/http2-spec/#GOAWAY):

```
There is an inherent race condition between an endpoint starting new streams and the remote sending a GOAWAY frame. 
To deal with this case, the GOAWAY contains the stream identifier of the last peer-initiated stream that 
was or might be processed on the sending endpoint in this connection. 
For instance, if the server sends a GOAWAY frame, the identified stream is the highest-numbered stream initiated by the client.
```

That is why "last accepted stream/request id" with GOAWAY is important.

**Comment by stevegury** [comment]
@NiteshKant you're right, GOAWAY is stricly better.

**Comment by tmontgomery** [comment]
SETUP_ERROR has been removed and replaced with a single ERROR frame that handles all of this. Error codes updated.

**Comment by NiteshKant** [comment]
Are we intending to use CONNECTION_ERROR as something analogous to GOAWAY?

**Comment by benjchristensen** [comment]
We have added a CONNECTION_ERROR to the protocol for optimistic delivery of a connection failure.

**Comment by NiteshKant** [comment]
Reopening this as per offline and related discussions in #92 

The general agreement on this issue was that an ERROR frame on stream 0 and no further leases is equivalent to HTTP/2 GOAWAY frame. However, as it turns out, leases are optional, so a client may not be honoring leases. In such a scenario, an ERROR frame on stream id 0 does not indicate a graceful shutdown and is ambiguous.

I propose we introduce something like GOAWAY.

**Comment by NiteshKant** [comment]
\cc @stevegury @yschimke @yunong

**Comment by stevegury** [comment]
👍 
it simplifies the code for dealing with streamIds exhaustion, and it can also be used for properly close a server (with draining of clients.)

**Comment by benjchristensen** [comment]
cc @tmontgomery

**Comment by tmontgomery** [comment]
I agree that GOAWAY might be simpler for handling stream ID exhaustion. But it does raise a few/several points.

- What semantics should be assumed for existing streams when GOAWAY is seen? Implicit end of stream?
- What about outstanding requests? Error indicated to the application?

We should be able to come up with something simple here, though.

**Comment by stevegury** [comment]
Below is what I believe we need for reactivesocket.

> What semantics should be assumed for existing streams when GOAWAY is seen? Implicit end of stream?

All the active streams stay active.

> What about outstanding requests? Error indicated to the application?

All outstanding requests are implicitly error/refused via the GOAWAY frame. There is no ambiguity because the GOAWAY frame contains the streamId of the last accepted stream.

**Comment by yschimke** [comment]
BTW for context, I forgot about this and dreamed up a metadataPush carrying a DRAIN message which is essentially the same thing.  So I'm a big +1 for GOAWAY with meaning of stop sending new requests and optionally you have X seconds before the connection will be closed.

**Comment by tmontgomery** [comment]
From a discussion today:

GOAWAY needs to indicate a temporary condition (retry OK) vs. a permanent one if resumability is supported.

**Comment by NiteshKant** [comment]
@tmontgomery @benjchristensen anyone of you wants to add this to the spec?

**Comment by benjchristensen** [comment]
@tmontgomery should this be put in before we call the spec 1.0?

Should this be a new frame type, or just another error code, such as CONNECTION_GOAWAY?

**Comment by tmontgomery** [comment]
I would be OK with 1.0 using the error codes for CONNECTION_ERROR for now to get the semantics when needed. Would kind of like more implementation experience to see if we need more of a controlled GOAWAY/FIN sequence than the error conditions.

**Comment by benjchristensen** [comment]
Removing 1.0 milestone then.

**Comment by NiteshKant** [comment]
>>> I would be OK with 1.0 using the error codes for CONNECTION_ERROR for now to get the semantics when needed.

I agree that we don't need to have a different frame type for 1.0 (probably never) but I think we should __at least do one__ of the following:
- Mention that all connection errors require draining i.e. error all streams before closing connection as a response to the error.
- Define a new error code for GRACEFUL_SHUTDOWN

**Comment by tmontgomery** [comment]
@NiteshKant can you suggest some wording here?

**Comment by NiteshKant** [comment]
@tmontgomery the question is do we update the `CONNECTION_ERROR` description to mention that all close should be graceful (wait for all active streams to terminate) or we introduce another error code like `GRACEFUL_SHUTDOWN` and keep the current `CONNECTION_ERROR` as immediate shutdown. I prefer the latter, WDYT?

**Comment by tmontgomery** [comment]
I think I prefer the latter, but not sure exactly what that would look like....

**Comment by NiteshKant** [comment]
@tmontgomery thanks, I will send a PR.

**Comment by benjchristensen** [comment]
This was merged in #148
